File: src/ios/AirbridgeEvent.m
//
//  AirbridgeEvent.m
//  AirbridgeCO
//
//  Created by WOF on 10/05/2019.
//  Copyright © 2019 ab180. All rights reserved.
//

#import "AirbridgeEvent.h"

#import "ACOMacro.h"

// Airbridge
#import <AirBridge/AirBridge.h>
#import <AirBridge/ABInAppEvent.h>
#import <AirBridge/ABCategory.h>

// AB
#import "ACOGet.h"

@implementation AirbridgeEvent

//
// cordova method
//

/**
 * send event
 * <p>
 *
 * @param       command
 *                  CDVInvokedUrlCommand
 */
- (void)trackEvent:(CDVInvokedUrlCommand*)command {
    NSString* category = [ACOGet type:NSString.class array:command.arguments index:0];
    if (category == nil) {
        NSLog(@"[Airbridge CO iOS] fail to send custom event - category not found");
        cordovaFail(self, command, @"category not found");
        return;
    }
    NSDictionary* option = [ACOGet type:NSDictionary.class array:command.arguments index:1];

    ABInAppEvent* event = [[ABInAppEvent alloc] init];

    addOptionToEvent(event, option);
    [event setCategory:category];
    
    [event send];

    cordovaSuccess(self, command);
}

//
// private - tool
//

/**
 * Add option infomation to event
 * @discussion  add nil when some data has unmatch type
 * @param       event
 *                  target event
 * @param       option
 *                  dictionary which has option infomation
 */
static void addOptionToEvent(ABInAppEvent* event, NSDictionary* __nullable option) {
    if (option == nil) return;
    
    NSString* action = [ACOGet type:NSString.class dictionary:option key:@"action"];
    NSString* label = [ACOGet type:NSString.class dictionary:option key:@"label"];
    NSNumber* value = [ACOGet type:NSNumber.class dictionary:option key:@"value"];
    NSDictionary* attributes = [ACOGet type:NSDictionary.class dictionary:option key:@"customAttributes"];
    NSDictionary* semantics = [ACOGet type:NSDictionary.class dictionary:option key:@"semanticAttributes"];
    
    [event setAction:action];
    [event setLabel:label];
    [event setValue:value];
    [event setCustoms:attributes];
    [event setSemantics:semantics];
}

@end

File: src/android/AirbridgeEvent.java
package co.ab180.airbridge.cordova;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import co.ab180.airbridge.Airbridge;
import co.ab180.airbridge.event.Event;
import co.ab180.airbridge.event.StandardEventCategory;

public class AirbridgeEvent extends CordovaPlugin {

    //
    // cordova method
    //

    public void trackEvent(@Nullable String category, @Nullable JSONObject option) {
        if (category == null) {
            return;
        }
        
        Event event = new Event(category);
        addOptionToEvent(event, option);
        Airbridge.trackEvent(event);
    }

    //
    // private - tool
    //

    /**
     * Add option infomation to event
     * <p>
     * add null when some data has unmatch type
     * @param       event
     *                  target event
     * @param       option
     *                  json-object which has option infomation
     */
    private void addOptionToEvent(@Nonnull Event event, @Nullable JSONObject option) {
        if (option == null) return;

        String action = Get.type(String.class, option, "action");
        String label = Get.type(String.class, option, "label");
        Float value = Get.type(Float.class, Get.type(Number.class, option, "value"));
        Map<String, Object> attributes = convertMap(Get.type(JSONObject.class, option, "customAttributes"));
        Map<String, Object> semantics = convertMap(Get.type(JSONObject.class, option, "semanticAttributes"));

        event.setAction(action);
        event.setLabel(label);
        event.setValue(value);
        event.setCustomAttributes(attributes);
        event.setSemanticAttributes(semantics);
    }

    /**
     * convert json-object to Map
     * <p>
     * return null when input is null
     * @param       input
     *                  json-object
     * @return      Map
     */
    @Nullable
    private Map convertMap(@Nullable JSONObject input) {
        if (input == null) return null;

        Map<String, Object> map = new HashMap<>();

        Iterator<String> it = input.keys();
        while(it.hasNext()) {
            String key = it.next();
            Object value = input.opt(key);

            if (value instanceof JSONObject) {
                value = convertMap((JSONObject) value);
            } else if (value instanceof JSONArray) {
                value = convertList((JSONArray) value);
            }

            if (key != null && value != null) {

                map.put(key, value);
            }
        }

        return map;
    }

    /**
     * convert json-array to List
     * <p>
     * return null when input is null
     * @param       input
     *                  json-array
     * @return      List
     */
    @Nullable
    private List convertList(@Nullable JSONArray input) {
        if (input == null) return null;

        List<Object> list = new ArrayList<>();

        for (int i=0; i<input.length(); i++) {
            Object value = input.opt(i);

            if (value instanceof JSONObject) {
                value = convertMap((JSONObject) value);
            } else if (value instanceof JSONArray) {
                value = convertList((JSONArray) value);
            }

            if (value != null) {
                list.add(value);
            }
        }

        return list;
    }

    //
    // override
    //

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {
        try {
            return execute0(action, args, callbackContext);
        } catch(Exception e) {
            callbackContext.error(e.getMessage());
            return false;
        }
    }

    public boolean execute0(String action, JSONArray args, CallbackContext callbackContext) {
        switch (action) {
            case "trackEvent":
                trackEvent(args.optString(0), args.optJSONObject(1));
                break;
            default:
                return false;
        }
        callbackContext.success();

        return true;
    }
}
